dialogue-test-common/src/main/java/com/palantir/dialogue/AbstractHandshakeTimeoutTest.java

```diff
@@ -0,0 +1,203 @@
+import org.xnio.XnioWorker;
+
+public abstract class AbstractHandshakeTimeoutTest {
+
+    private static final Request request = Request.builder()
+            .body(new RequestBody() {
+                @Override
+                public void writeTo(OutputStream output) throws IOException {
+                    output.write("Hello, World".getBytes(StandardCharsets.UTF_8));
+                }
+
+                @Override
+                public String contentType() {
+                    return "text/plain";
+                }
+
+                @Override
+                public boolean repeatable() {
+                    return true;
+                }
+
+                @Override
+                public void close() {}
+            })
+            .build();
+
+    protected abstract Channel create(ClientConfiguration config);
+
+    private static int getPort(Undertow undertow) {
+        return ((InetSocketAddress)
+                        Iterables.getOnlyElement(undertow.getListenerInfo()).getAddress())
+                .getPort();
+    }
+
+    @Test
+    public void testHandshakeTimeout() throws Exception {
+        SSLContext sslContext = SslSocketFactories.createSslContext(TestConfigurations.SSL_CONFIG);
+        Xnio xnio = Xnio.getInstance(Undertow.class.getClassLoader());
+        DelayingNextTaskExecutorService executor = new DelayingNextTaskExecutorService(Executors.newCachedThreadPool());
+        XnioWorker worker = xnio.createWorkerBuilder()
+                .setWorkerIoThreads(1)
+                .setExternalExecutorService(executor)
+                .build();
+        Undertow server = Undertow.builder()
+                .setWorker(worker)
+                .setHandler(ResponseCodeHandler.HANDLE_200)
+                .addHttpsListener(0, null, sslContext)
+                .build();
+        server.start();
+        try {
+            int serverPort = getPort(server);
+            ClientConfiguration noRetryConfig = ClientConfiguration.builder()
+                    .from(TestConfigurations.create("https://localhost:" + serverPort))
+                    .connectTimeout(Duration.ofMillis(500))
+                    .readTimeout(Duration.ofMillis(500))
+                    .writeTimeout(Duration.ofMillis(500))
+                    .maxNumRetries(0)
+                    .build();
+            Channel noRetryChannel = create(noRetryConfig);
+
+            executor.delayNextTask(Duration.ofSeconds(1));
+
+            assertThatThrownBy(noRetryChannel.execute(TestEndpoint.POST, request)::get)
+                    .getCause()
+                    .satisfies(cause -> assertThat(cause)
+                            .isInstanceOf(SafeLoggable.class)
+                            .isInstanceOf(IOException.class)
+                            .isNotInstanceOf(SocketTimeoutException.class)
+                            .extracting(value -> value.getClass().getSimpleName())
+                            .isEqualTo("SafeConnectTimeoutException"));
```

Comment:
need to do all of this since `SafeConnectTimeoutException` is packagePrivate

Reply:
good call, originally it was nested+private!

Reply:
eh disregard, these are in different modules. Since we only really care about hc at this point I can move the test and get rid of the abstract-ion.